weldx/measurement.py

```diff
@@ -55,6 +70,332 @@ class Source:
+            source, if no transformation was added to the chain) is used.
+
+        """
+        input_signal_source = self._check_and_get_node_name(input_signal_source)
+
+        self._add_signal(
+            node_id=name, signal_type=output_signal_type, unit=output_signal_unit
+        )
+        self._graph.add_edge(input_signal_source, name, error=error, func=func)
+
+    def add_signal_data(self, name: str, data: "TimeSeries", signal_source: str = None):
+        """Add data to a signal.
+
+        Parameters
+        ----------
+        name :
+            Name of the data
+        data :
+            The data that should be added
+        signal_source :
+            The source of the signal that the data should be attached to.
+            If `None` is provided, the name of the last added transformation (or the
+            source, if no transformation was added to the chain) is used.
+
+        """
+        self._raise_if_data_exist(name)
+
+        signal_source = self._check_and_get_node_name(signal_source)
+        signal = self._graph.nodes[signal_source]
+
+        signal["data_name"] = name
+        signal["data"] = data
+
+    def attach_transformation(
+        self, transformation: SignalTransformation, input_signal_source: str = None
+    ):
+        """Add a transformation from an `SignalTransformation` instance.
+
+        Parameters
+        ----------
+        transformation :
+            The class containing the transformation data.
+        input_signal_source :
+            The source of the signal that should be used as input of the transformation.
+            If `None` is provided, the name of the last added transformation (or the
+            source, if no transformation was added to the chain) is used.
+
+        """
+        if input_signal_source is None:
+            input_signal_source = self._prev_added_signal
+
+        source_node = self._graph.nodes[input_signal_source]
+        if (
+            transformation.input_signal.signal_type != source_node["signal_type"]
+            or transformation.input_signal.unit != source_node["unit"]
+        ):
+            raise ValueError(
+                f"The provided transformations input signal is incompatible to the "
+                f"output signal of {input_signal_source}:\n"
+                f"transformation: {transformation.input_signal.signal_type} in ["
+                f"{transformation.input_signal.unit}]\n"
+                f"output signal : {source_node['signal_type']} in ["
+                f"{source_node['unit']}]"
+            )
+
+        self.add_transformation(
+            transformation.name,
+            transformation.error,
+            transformation.output_signal.signal_type,
+            transformation.output_signal.unit,
+            transformation.func,
+            input_signal_source,
+        )
+
+    def get_data(self, name: str) -> xr.DataArray:
```

Comment:
I think it would be ok if data itself does not have a name but is simply attached to a signal, so either a signal has data or not and just access it via the signal. What do you think?

Reply:
What if a signal has multiple data blocks? For example, expected and measured? Also, the signal also has no name, so we have to use the name of the signal source.

Reply:
I don't think that fits the current scope of the measurement chain which should describe exactly one measurement.
Not sure how e.g. an expected value would fit into that, maybe that should it described in it's own measurement chain? Or just an external signal?

Reply:
We can remove the name. I'll change the interfaces accordingly